src/lsm/table_memory.zig

```diff
@@ -136,23 +148,102 @@ pub fn TableMemoryType(comptime Table: type) type {
             table.* = .{
                 .values = table.values,
                 .value_context = .{},
-                .mutability = .mutable,
+                .mutability = .{ .mutable = .{} },
                 .name = table.name,
             };
         }
 
         pub fn sort(table: *TableMemory) void {
+            assert(table.mutability == .mutable);
+
             if (!table.value_context.sorted) {
-                std.mem.sort(
-                    Value,
-                    table.values_used(),
-                    {},
-                    sort_values_by_key_in_ascending_order,
-                );
+                table.sort_suffix_from_offset(0);
                 table.value_context.sorted = true;
             }
         }
 
+        pub fn sort_suffix(table: *TableMemory) void {
+            assert(table.mutability == .mutable);
+            assert(table.mutability.mutable.suffix_offset <= table.count());
+
+            table.sort_suffix_from_offset(table.mutability.mutable.suffix_offset);
+
+            assert(table.mutability.mutable.suffix_offset == table.count());
+        }
+
+        fn sort_suffix_from_offset(table: *TableMemory, offset: u32) void {
+            assert(table.mutability == .mutable);
+            assert(offset == table.mutability.mutable.suffix_offset or offset == 0);
+            assert(offset <= table.count());
+
+            std.mem.sort(
+                Value,
+                table.values_used()[offset..],
+                {},
+                sort_values_by_key_in_ascending_order,
+            );
+
+            // Merge values with identical keys (last one wins) and collapse tombstones for
+            // secondary indexes.
+            const source_count: u32 = table.count();
+            var source_index: u32 = offset;
+            var target_index: u32 = offset;
+            while (source_index < source_count) {
+                table.values[target_index] = table.values[source_index];
```

Comment:
This reassignment bugs me a bit, for the case where target_index == source_index. I don't know if the semantics here is of `memcpy` or of `copyForward`?

Not sure if an extra `if source_index == target_index` is warranted here.

But maybe we can do better than an if? WDYT about adding fast-path for finding the _first_ index we need to skip? So that, by the time you enter this loop, you already have `source_index < target_index`? 

```
const first_mismatch = for(source_index..source_count) {
   // just read array elements here
} orelse return; // fast path, nothing to collapse

const source_index = first_mismatch + 1;
const target_index = first_mismatch;
assert(target_index < source_index);
```